### PR TITLE
Add Fillet Away + Belly Drum tweaks

### LINKED ISSUE #2499: MOVE_FILLET_AWAY Effect

### PR DESCRIPTION
**Pull request by kittenchilly**
## Description
Adds Fillet Away + tests. Also tweaks Belly Drum (and adds tests for it) so we can reuse code between it and Fillet Away.

## Issue(s) that this PR fixes
Closes #2499

## **Discord contact info**
kittenchilly

### CODE REVIEW

File: test/battle/move_effect/fillet_away.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        TURN { MOVE(player, MOVE_FILLET_AWAY); }
+    } SCENE {
+        MESSAGE("But it failed!");
+        NONE_OF {
+            ANIMATION(ANIM_TYPE_MOVE, MOVE_FILLET_AWAY, player);
+            HP_BAR(player);
+            ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player);
+        }
+    }
+}
```
Comment (Bassoonian) [inline test/battle/move_effect/fillet_away.c:58]: Missing newline

File: test/battle/move_effect/belly_drum.c [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        MESSAGE("Wobbuffet's Attack sharply rose!");
+
+        MESSAGE("But it failed!");
+        NONE_OF {
+            ANIMATION(ANIM_TYPE_MOVE, MOVE_BELLY_DRUM, player);
+            ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player);
+            HP_BAR(player);
+        }
+    }
+}
```
Comment (Bassoonian) [inline test/battle/move_effect/belly_drum.c:91]: Missing newline

**Review by Bassoonian — CHANGES_REQUESTED** [review]

File: data/battle_scripts_1.s [hunk 3]
```diff
@@ -442,6 +442,21 @@ gBattleScriptsForMoveEffects::
 	.4byte BattleScript_EffectMaxMove                 @ EFFECT_MAX_MOVE
 	.4byte BattleScript_EffectGlaiveRush              @ EFFECT_GLAIVE_RUSH
 	.4byte BattleScript_EffectBrickBreak              @ EFFECT_RAGING_BULL
+	.4byte BattleScript_EffectFilletAway              @ EFFECT_FILLET_AWAY
+
+BattleScript_EffectFilletAway:
```
Comment (Skeli789) [inline data/battle_scripts_1.s:447]: The correct order of Fillet Away is first boost stats, then cut HP. Also, the current way you have it, the attack animation would still play even if the stats can't be lowered. I recommend keeping the jump to Shell Smash, and then at the end of the Shell Smash boosts, do something like `jumpifmoveeffect EFFECT_FILLETAWAY BattleScript_FilletAwayCutHP`.

File: data/battle_scripts_1.s [hunk 4]
```diff
@@ -5272,12 +5287,16 @@ BattleScript_EffectBellyDrum::
...
-	maxattackhalvehp BattleScript_ButItFailed
+	jumpifstat BS_ATTACKER, CMP_EQUAL, STAT_ATK, MAX_STAT_STAGE, BattleScript_ButItFailed
+	halvehp BattleScript_ButItFailed
 	orword gHitMarker, HITMARKER_IGNORE_SUBSTITUTE
 	attackanimation
 	waitanimation
 	healthbarupdate BS_ATTACKER
 	datahpupdate BS_ATTACKER
+	playstatchangeanimation BS_ATTACKER, BIT_ATK, STAT_CHANGE_BY_TWO
+	setstatchanger STAT_ATK, 12, FALSE
```
Comment (Skeli789) [inline data/battle_scripts_1.s:5298]: Is there no constant like `STAT_MAX` instead of `12`?

File: data/battle_scripts_1.s [hunk 5]
```diff
@@ -5272,12 +5287,16 @@ BattleScript_EffectBellyDrum::
 	attackcanceler
```
Comment (Skeli789) [inline data/battle_scripts_1.s:5287]: You should test this script with Contrary to ensure you didn't break anything.

**Review by Skeli789 — CHANGES_REQUESTED** [review]

File: data/battle_scripts_1.s (hunk 3 above)
Comment (kittenchilly) [reply data/battle_scripts_1.s:447]: Is there a jumpifmoveeffect? If we have to do this then it'd be easier to just not jump to shell smash code anymore and do the stat changes here.

Comment (Skeli789) [reply data/battle_scripts_1.s:447]: By default I don't believe that battle script command exists, but it's not so hard to implement

**Comment by kittenchilly** [comment]
Should be reviewable again.

File: data/battle_scripts_1.s [hunk 6]
```diff
@@ -440,6 +440,42 @@ gBattleScriptsForMoveEffects::
...
+BattleScript_FilletAwayTrySpeed:
+	setstatchanger STAT_SPEED, 2, FALSE
+	statbuffchange MOVE_EFFECT_AFFECTS_USER | STAT_CHANGE_ALLOW_PTR, BattleScript_FilletAwayEnd
+	jumpifbyte CMP_EQUAL, cMULTISTRING_CHOOSER, B_MSG_STAT_WONT_INCREASE, BattleScript_FilletAwayEnd
+	printfromtable gStatUpStringIds
+	waitmessage B_WAIT_TIME_LONG
+BattleScript_FilletAwayEnd:
+	healthbarupdate BS_ATTACKER
+	datahpupdate BS_ATTACKER
+	goto BattleScript_MoveEnd
```
Comment (AlexOn1ine) [inline data/battle_scripts_1.s:478]: When 2 stats are maxed out the health isn't halved.

File: test/battle/ability/contrary.c [hunk 7]
```diff
@@ -187,3 +187,40 @@ SINGLE_BATTLE_TEST("Contrary raises a stat after using a move which would normal
...
+        OPPONENT(SPECIES_SPINDA) { Ability(ability); }
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_TACKLE); }
+        TURN { MOVE(opponent, MOVE_BELLY_DRUM); }
+        TURN { MOVE(opponent, MOVE_TACKLE); }
+    } SCENE {
+        MESSAGE("Foe Spinda used Tackle!");
+        HP_BAR(player, captureDamage: &results[i].damageBefore);
+
+        //MESSAGE("Foe Spinda used Belly Drum!");
```
Comment (AlexOn1ine) [inline test/battle/ability/contrary.c:208]: Remove the commented out code in this test if it's not needed anymore.

**Review by AlexOn1ine — CHANGES_REQUESTED** [review]

File: data/battle_scripts_1.s (hunk 6 above)
Comment (kittenchilly) [reply data/battle_scripts_1.s:478]: Which 2 stats were maxed out to cause this?

Comment (AlexOn1ine) [reply data/battle_scripts_1.s:478]: special attack and speed.

Comment (kittenchilly) [reply data/battle_scripts_1.s:478]: Thats weird, it should go to BattleScript_FilletAwayTryAttack which runs halvehp...

File: data/battle_scripts_1.s [hunk 8]
```diff
@@ -440,6 +440,42 @@ gBattleScriptsForMoveEffects::
...
+	ppreduce
+	jumpifstat BS_ATTACKER, CMP_LESS_THAN, STAT_ATK, MAX_STAT_STAGE, BattleScript_FilletAwayTryAttack
+	jumpifstat BS_ATTACKER, CMP_LESS_THAN, STAT_SPATK, MAX_STAT_STAGE, BattleScript_FilletAwayTryAttack
+	jumpifstat BS_ATTACKER, CMP_EQUAL, STAT_SPEED, MAX_STAT_STAGE, BattleScript_ButItFailed
+BattleScript_FilletAwayTryAttack:
+	halvehp BattleScript_ButItFailed
+	attackanimation
+	waitanimation
+	setbyte sSTAT_ANIM_PLAYED, FALSE
+	playstatchangeanimation BS_ATTACKER, BIT_SPATK | BIT_ATK | BIT_SPEED, STAT_CHANGE_BY_TWO
```
Comment (Skeli789) [inline data/battle_scripts_1.s:457]: Minor aesthetic change but change the order to match the order of boosts
```suggestion
	playstatchangeanimation BS_ATTACKER, BIT_ATK | BIT_SPATK | BIT_SPEED, STAT_CHANGE_BY_TWO
```

File: data/battle_scripts_1.s [hunk 9]
```diff
@@ -440,6 +440,42 @@ gBattleScriptsForMoveEffects::
...
+	setstatchanger STAT_ATK, 2, FALSE
+	statbuffchange MOVE_EFFECT_AFFECTS_USER | STAT_CHANGE_ALLOW_PTR, BattleScript_FilletAwayTrySpAtk
+	jumpifbyte CMP_EQUAL, cMULTISTRING_CHOOSER, B_MSG_STAT_WONT_INCREASE, BattleScript_FilletAwayTrySpAtk
```
Comment (Skeli789) [inline data/battle_scripts_1.s:460]: Just please confirm in-game it's not supposed to play an "Attack won't go higher!" string when Attack is maxed but other stats are not.

**Review by Skeli789 — CHANGES_REQUESTED** [review]
Also missing AI logic

File: data/battle_scripts_1.s (hunk 9 above)
Comment (kittenchilly) [reply data/battle_scripts_1.s:460]: How would I fix this? I based it off of code from shell smash, meaning shell smash should have the same issue.

File: data/battle_scripts_1.s (hunk 6 above)
Comment (kittenchilly) [reply data/battle_scripts_1.s]: I'm really stumped on this.

File: data/battle_scripts_1.s (hunk 9 above)
Comment (Skeli789) [reply data/battle_scripts_1.s:460]: You can just remove the line if that ends up being the proper behaviour

Comment (kittenchilly) [reply data/battle_scripts_1.s]: Alright, just tested it and I confirmed it does actually play the won't go higher text. Fixed.

File: data/battle_scripts_1.s [hunk 10]
```diff
@@ -440,6 +440,41 @@ gBattleScriptsForMoveEffects::
...
+BattleScript_FilletAwayTrySpAtk::
+	setstatchanger STAT_SPATK, 2, FALSE
+	statbuffchange MOVE_EFFECT_AFFECTS_USER | STAT_CHANGE_ALLOW_PTR, BattleScript_FilletAwayTrySpeed
+	printfromtable gStatUpStringIds
+	waitmessage B_WAIT_TIME_LONG
+BattleScript_FilletAwayTrySpeed::
+	setstatchanger STAT_SPEED, 2, FALSE
+	statbuffchange MOVE_EFFECT_AFFECTS_USER | STAT_CHANGE_ALLOW_PTR, BattleScript_FilletAwayEnd
+	printfromtable gStatUpStringIds
+	waitmessage B_WAIT_TIME_LONG
```
Comment (AlexOn1ine) [inline data/battle_scripts_1.s:472]: Wrong speed message is displayed when when both Special Attack and Speed are maxed out

**Review by AlexOn1ine — CHANGES_REQUESTED** [review]

**Review by AlexOn1ine — APPROVED** [review]

**Comment by AlexOn1ine** [comment]
@Skeli789 Thank you for your help as always!

File: src/battle_ai_main.c [hunk 11]
```diff
@@ -1699,6 +1699,7 @@ static s32 AI_CheckBadMove(u32 battlerAtk, u32 battlerDef, u32 move, s32 score)
                 ADJUST_SCORE(-6);
             break;
         case EFFECT_BELLY_DRUM:
+        case EFFECT_FILLET_AWAY:
```
Comment (Skeli789) [inline src/battle_ai_main.c:1702]: Both Belly Drum and Fillet Away are missing logic for when the stats are already maxed

**Pull request merged**